File: app/src/main/java/de/ironjan/mensaupb/stw/rest_api/Allergen.java
package de.ironjan.mensaupb.stw.rest_api;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

import org.slf4j.LoggerFactory;

import de.ironjan.mensaupb.R;
import de.ironjan.mensaupb.stw.rest_api.deserializer.AllergenDeserializer;

/**
 * Class that represents Allergens and Additionals.
 */
@SuppressWarnings("MagicNumber")
@JsonDeserialize(using = AllergenDeserializer.class)
public enum Allergen implements Comparable<Allergen> {
    UNKNOWN(0, "", R.string.empty),
    COLORED(1, Constants.COLORED, R.string.colored),
    CONSERVED(2, Constants.CONSERVED, R.string.conserved),
    ANTIOXIDANTS(3, Constants.ANTIOXIDANTS, R.string.antioxidants),
    FLAVOR_ENHANCERS(4, Constants.FLAVOR_ENHANCERS, R.string.tasteEnhancer),
    PHOSPHAT(5, Constants.PHOSPHAT, R.string.phosphate),
    SULFURATED(6, Constants.SULFURATED, R.string.sulfured),
    WAXED(7, Constants.WAXED, R.string.waxed),
    BLACKENED(8, Constants.BLACKENED, R.string.blackened),
    SWEETENER(9, Constants.SWEETENER, R.string.sweetener),
    PHENYLALANINE(10, Constants.PHENYLALANINE, R.string.phenylalanine),
    TAURINE(11, Constants.TAURINE, R.string.taurine),
    NITRATE_SALT(12, Constants.NITRATE_SALT, R.string.nitrate_salt),
    COFFEINE(13, Constants.COFFEINE, R.string.caffeine),
    QUININE(14, Constants.QUININE, R.string.quinine),
    LACTOPROTEIN(15, Constants.LACTOPROTEIN, R.string.lacto_protein),
    CRUSTACEAN(16, Constants.CRUSTACEAN, R.string.crustacean),
    EGGS(17, Constants.EGGS, R.string.eggs),
    FISH(18, Constants.FISH, R.string.fish),
    SOYA(19, Constants.SOYA, R.string.soy),
    LACTOSE(20, Constants.LACTOSE, R.string.milk),
    NUTS(21, Constants.NUTS, R.string.nuts),
    CELERIAC(22, Constants.CELERIAC, R.string.celeriac),
    MUSTARD(23, Constants.MUSTARD, R.string.mustard),
    SESAME(24, Constants.SESAME, R.string.sesame),
    SULFITES(25, Constants.SULFITES, R.string.sulfates),
    MOLLUSKS(26, Constants.MOLLUSKS, R.string.mollusks),
    LUPINE(27, Constants.LUPINE, R.string.lupines),
    GLUTEN(28, Constants.GLUTEN, R.string.gluten),
    PEANUTS(29, Constants.PEANUTS, R.string.peanuts);

    private final int ordinal;
    private final String type;
    private final int stringResourceID;

    Allergen(int ordinal, String type, int stringResourceID) {
        this.ordinal = ordinal;
        this.type = type;
        this.stringResourceID = stringResourceID;
    }

    /**
     * Converts a string to a Allergen
     *
     * @param string the string
     * @return the Allergen which is textually represented as String, if known. Else "Unknown".
     */
    public static Allergen fromString(String string) {
        switch (string) {
            case Constants.COLORED:
                return COLORED;
            case Constants.CONSERVED:
                return CONSERVED;
            case Constants.ANTIOXIDANTS:
                return ANTIOXIDANTS;
            case Constants.FLAVOR_ENHANCERS:
                return FLAVOR_ENHANCERS;
            case Constants.PHOSPHAT:
                return PHOSPHAT;
            case Constants.SULFURATED:
                return SULFURATED;
            case Constants.WAXED:
                return WAXED;
            case Constants.BLACKENED:
                return BLACKENED;
            case Constants.SWEETENER:
                return SWEETENER;
            case Constants.PHENYLALANINE:
                return PHENYLALANINE;
            case Constants.TAURINE:
                return TAURINE;
            case Constants.NITRATE_SALT:
                return NITRATE_SALT;
            case Constants.COFFEINE:
                return COFFEINE;
            case Constants.QUININE:
                return QUININE;
            case Constants.LACTOPROTEIN:
                return LACTOPROTEIN;
            case Constants.CRUSTACEAN:
                return CRUSTACEAN;
            case Constants.EGGS:
                return EGGS;
            case Constants.FISH:
                return FISH;
            case Constants.SOYA:
                return SOYA;
            case Constants.LACTOSE:
                return LACTOSE;
            case Constants.NUTS:
                return NUTS;
            case Constants.CELERIAC:
                return CELERIAC;
            case Constants.MUSTARD:
                return MUSTARD;
            case Constants.SESAME:
                return SESAME;
            case Constants.SULFITES:
                return SULFITES;
            case Constants.MOLLUSKS:
                return MOLLUSKS;
            case Constants.LUPINE:
                return LUPINE;
            case Constants.GLUTEN:
                return GLUTEN;
            case Constants.PEANUTS:
                return PEANUTS;
        }

        LoggerFactory.getLogger(Allergen.class).debug("Requested unknown value: " + string);
        return UNKNOWN;
    }

    @Override
    public String toString() {
        return type;
    }

    public String getType() {
        return type;
    }

    public int getOrdinal() {
        return ordinal;
    }

    public int getStringId() {
        return stringResourceID;
    }


    private static class Constants {
        public static final String COLORED = "1";
        public static final String CONSERVED = "2";
        public static final String ANTIOXIDANTS = "3";
        public static final String FLAVOR_ENHANCERS = "4";
        public static final String PHOSPHAT = "5";
        public static final String SULFURATED = "6";
        public static final String WAXED = "7";
        public static final String BLACKENED = "8";
        public static final String SWEETENER = "9";
        public static final String PHENYLALANINE = "10";
        public static final String TAURINE = "11";
        public static final String NITRATE_SALT = "12";
        public static final String COFFEINE = "13";
        public static final String QUININE = "14";
        public static final String LACTOPROTEIN = "15";
        public static final String CRUSTACEAN = "A2";
        public static final String EGGS = "A3";
        public static final String FISH = "A4";
        public static final String SOYA = "A6";
        public static final String LACTOSE = "A7";
        public static final String NUTS = "A8";
        public static final String CELERIAC = "A9";
        public static final String MUSTARD = "A10";
        public static final String SESAME = "A11";
        public static final String SULFITES = "A12";
        public static final String LUPINE = "A13";
        public static final String MOLLUSKS = "A14";
        public static final String GLUTEN = "A1";
        public static final String PEANUTS = "A5";
    }


}

File: app/src/main/kotlin/de/ironjan/mensaupb/api/model/Allergen.kt
package de.ironjan.mensaupb.api.model

import de.ironjan.mensaupb.R
import org.slf4j.LoggerFactory

/**
 * Class that represents Allergens and Additionals.
 */
enum class Allergen(val type: String, val stringId: Int) : Comparable<Allergen> {
    UNKNOWN("", R.string.empty),
    COLORED(Constants.COLORED, R.string.colored),
    CONSERVED(Constants.CONSERVED, R.string.conserved),
    ANTIOXIDANTS(Constants.ANTIOXIDANTS, R.string.antioxidants),
    FLAVOR_ENHANCERS(Constants.FLAVOR_ENHANCERS, R.string.tasteEnhancer),
    PHOSPHAT(Constants.PHOSPHAT, R.string.phosphate),
    SULFURATED(Constants.SULFURATED, R.string.sulfured),
    WAXED(Constants.WAXED, R.string.waxed),
    BLACKENED(Constants.BLACKENED, R.string.blackened),
    SWEETENER(Constants.SWEETENER, R.string.sweetener),
    PHENYLALANINE(Constants.PHENYLALANINE, R.string.phenylalanine),
    TAURINE(Constants.TAURINE, R.string.taurine),
    NITRATE_SALT(Constants.NITRATE_SALT, R.string.nitrate_salt),
    COFFEINE(Constants.COFFEINE, R.string.caffeine),
    QUININE(Constants.QUININE, R.string.quinine),
    LACTOPROTEIN(Constants.LACTOPROTEIN, R.string.lacto_protein),
    CRUSTACEAN(Constants.CRUSTACEAN, R.string.crustacean),
    EGGS(Constants.EGGS, R.string.eggs),
    FISH(Constants.FISH, R.string.fish),
    SOYA(Constants.SOYA, R.string.soy),
    LACTOSE(Constants.LACTOSE, R.string.milk),
    NUTS(Constants.NUTS, R.string.nuts),
    CELERIAC(Constants.CELERIAC, R.string.celeriac),
    MUSTARD(Constants.MUSTARD, R.string.mustard),
    SESAME(Constants.SESAME, R.string.sesame),
    SULFITES(Constants.SULFITES, R.string.sulfates),
    MOLLUSKS(Constants.MOLLUSKS, R.string.mollusks),
    LUPINE(Constants.LUPINE, R.string.lupines),
    GLUTEN(Constants.GLUTEN, R.string.gluten),
    PEANUTS(Constants.PEANUTS, R.string.peanuts);

    override fun toString(): String {
        return type
    }


    private object Constants {
        val COLORED = "1"
        val CONSERVED = "2"
        val ANTIOXIDANTS = "3"
        val FLAVOR_ENHANCERS = "4"
        val PHOSPHAT = "5"
        val SULFURATED = "6"
        val WAXED = "7"
        val BLACKENED = "8"
        val SWEETENER = "9"
        val PHENYLALANINE = "10"
        val TAURINE = "11"
        val NITRATE_SALT = "12"
        val COFFEINE = "13"
        val QUININE = "14"
        val LACTOPROTEIN = "15"
        val CRUSTACEAN = "A2"
        val EGGS = "A3"
        val FISH = "A4"
        val SOYA = "A6"
        val LACTOSE = "A7"
        val NUTS = "A8"
        val CELERIAC = "A9"
        val MUSTARD = "A10"
        val SESAME = "A11"
        val SULFITES = "A12"
        val LUPINE = "A13"
        val MOLLUSKS = "A14"
        val GLUTEN = "A1"
        val PEANUTS = "A5"
    }

    companion object {

        /**
         * Converts a string to a Allergen
         *
         * @param string the string
         * @return the Allergen which is textually represented as String, if known. Else "Unknown".
         */
        fun fromString(string: String): Allergen {
            when (string) {
                Constants.COLORED -> return COLORED
                Constants.CONSERVED -> return CONSERVED
                Constants.ANTIOXIDANTS -> return ANTIOXIDANTS
                Constants.FLAVOR_ENHANCERS -> return FLAVOR_ENHANCERS
                Constants.PHOSPHAT -> return PHOSPHAT
                Constants.SULFURATED -> return SULFURATED
                Constants.WAXED -> return WAXED
                Constants.BLACKENED -> return BLACKENED
                Constants.SWEETENER -> return SWEETENER
                Constants.PHENYLALANINE -> return PHENYLALANINE
                Constants.TAURINE -> return TAURINE
                Constants.NITRATE_SALT -> return NITRATE_SALT
                Constants.COFFEINE -> return COFFEINE
                Constants.QUININE -> return QUININE
                Constants.LACTOPROTEIN -> return LACTOPROTEIN
                Constants.CRUSTACEAN -> return CRUSTACEAN
                Constants.EGGS -> return EGGS
                Constants.FISH -> return FISH
                Constants.SOYA -> return SOYA
                Constants.LACTOSE -> return LACTOSE
                Constants.NUTS -> return NUTS
                Constants.CELERIAC -> return CELERIAC
                Constants.MUSTARD -> return MUSTARD
                Constants.SESAME -> return SESAME
                Constants.SULFITES -> return SULFITES
                Constants.MOLLUSKS -> return MOLLUSKS
                Constants.LUPINE -> return LUPINE
                Constants.GLUTEN -> return GLUTEN
                Constants.PEANUTS -> return PEANUTS
            }

            LoggerFactory.getLogger(Allergen::class.java).debug("Requested unknown value: $string")
            return UNKNOWN
        }
    }


}
